### ISSUE TITLE
implement bytemuck traits on some types to make it easy to integrate with opengl

### ISSUE THREAD

**Issue by coderedart** [emilk/egui#533]
most opengl crates take data as a &[u8], and in both glium and other implementations, people have to convert Vertex into their own struct that has bytemuck traits, or use **unsafe** casts :( . idk if converting between structs is a zero cost operation, but it will still be cool if egui support these traits. would be some nice QoL improvements for people integrating into opengl or such contexts. 

but it means adding a dependency and the traits themselves are unsafe , so , i would understand if the devs don't want to.

**Comment by t18b219k** [comment]
I think it awesome for wgpu or vulkano or other graphics backends

**Comment by t18b219k** [comment]
my backend have  unsafe cast for 
* epaint::Vertex
* &[epaint::Color32]

**Comment by coderedart** [comment]
pretty much all backends either use unsafe casting or convert to structs that implement bytemuck traits. that is why i would prefer if egui itself used those traits. that way we can also always be sure that the alignment with `#[repr(C)]` has been considered for safe casting to send to gpu memory.

**Comment by t18b219k** [comment]
#[repr(C)] attribute on Vertex there is no problem only we need  add Zeroable and Pod traits

**Comment by coderedart** [comment]
> #[repr(C)] attribute on Vertex there is no problem only we need add Zeroable and Pod traits

yes, i was referring to the `The type must not contain any padding bytes, either in the middle or on the end (eg: no #[repr(C)] struct Foo(u8, u16), which has padding in the middle, and also no #[repr(C)] struct Foo(u16, u8), which has padding on the end).` condition for bytemuck::Pod from the https://docs.rs/bytemuck/1.7.0/bytemuck/trait.Pod.html#safety . 

thus, if `epaint::Vertex` ever needs change (as its still not 1.0), then that condition has to be considered. that way users of egui can be sure that there's no padding or alignment issues, or if there are issues, they can be mentioned in the docs themselves

**Comment by coderedart** [comment]
I would just like to bump this again, as it seems to have gotten buried. implementing bytemuck traits could be nice Quality of Life improvement for the rendering crates. 

```rust
let vertices: Vec<Vertex> = mesh
                .vertices
                .iter()
                .map(|v| Vertex {
                    a_pos: [v.pos.x, v.pos.y],
                    a_tc: [v.uv.x, v.uv.y],
                    a_srgba: v.color.to_array(),
                })
                .collect();
``` this is the code in egui glium itself, where there's an unnecessary conversion to a local struct. 

and in the glow version (from the pull request) they use 
```rust
#[allow(unsafe_code)]
unsafe fn as_u8_slice<T>(s: &[T]) -> &[u8] {
    std::slice::from_raw_parts(s.as_ptr() as *const u8, s.len() * std::mem::size_of::<T>())
}``` to convert the vertices/indices. considering, Vertex struct is public and changing it is a breaking change, i think it is better to just implement bytemuck traits so that egui can itself verify if those structs are properly aligned.

**Comment by emilk** [comment]
Opt-in bytemuck support (behind a feature flag) seems reasonable to me
